File: src/test/java/no/nav/syfo/consumer/mq/InntektsmeldingConsumerIntegrassjonsTest.java
package no.nav.syfo.consumer.mq;

import no.nav.syfo.LocalApplication;
import no.nav.syfo.consumer.rest.aktor.AktorConsumer;
import no.nav.syfo.service.EksisterendeSakService;
import no.nav.syfo.consumer.ws.*;
import no.nav.syfo.domain.GeografiskTilknytningData;
import no.nav.syfo.domain.InngaaendeJournal;
import no.nav.syfo.domain.Periode;
import no.nav.syfo.repository.InntektsmeldingDAO;
import no.nav.syfo.repository.InntektsmeldingMeta;
import no.nav.syfo.util.Metrikk;
import no.nav.tjeneste.virksomhet.behandle.inngaaende.journal.v1.BehandleInngaaendeJournalV1;
import no.nav.tjeneste.virksomhet.journal.v2.HentDokumentDokumentIkkeFunnet;
import no.nav.tjeneste.virksomhet.journal.v2.HentDokumentSikkerhetsbegrensning;
import no.nav.tjeneste.virksomhet.journal.v2.JournalV2;
import no.nav.tjeneste.virksomhet.journal.v2.WSHentDokumentResponse;
import org.apache.activemq.command.ActiveMQTextMessage;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit4.SpringRunner;

import javax.inject.Inject;
import javax.jms.MessageNotWriteableException;
import java.time.LocalDate;
import java.util.Collections;
import java.util.List;

import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static no.nav.syfo.domain.InngaaendeJournal.MIDLERTIDIG;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = LocalApplication.class)
@TestPropertySource("classpath:application-test.properties")
@DirtiesContext
public class InntektsmeldingConsumerIntegrassjonsTest {

    @MockBean
    private InngaaendeJournalConsumer inngaaendeJournalConsumer;

    @MockBean
    private JournalV2 journalV2;

    @MockBean
    private AktorConsumer aktorConsumer;

    @MockBean
    private EksisterendeSakService eksisterendeSakService;

    @MockBean
    private BehandleSakConsumer behandleSakConsumer;

    @MockBean
    private BehandlendeEnhetConsumer behandlendeEnhetConsumer;

    @MockBean
    private OppgavebehandlingConsumer oppgavebehandlingConsumer;

    @MockBean
    private BehandleInngaaendeJournalV1 behandleInngaaendeJournalV1;

    @MockBean
    private Metrikk metrikk;

    @Inject
    private InntektsmeldingConsumer inntektsmeldingConsumer;

    @Inject
    private InntektsmeldingDAO inntektsmeldingDAO;

    @Inject
    private JdbcTemplate jdbcTemplate;

    private int saksIdteller = 0;

    @Before
    public void setup() {
        jdbcTemplate.update("DELETE FROM ARBEIDSGIVERPERIODE");
        jdbcTemplate.update("DELETE FROM INNTEKTSMELDING");
        saksIdteller = 0;

        InngaaendeJournal inngaaendeJournal = inngaaendeJournal("arkivId");

        when(behandleSakConsumer.opprettSak("fnr")).thenAnswer(invocation -> "saksId" + saksIdteller++);

        when(inngaaendeJournalConsumer.hentDokumentId("arkivId")).thenReturn(inngaaendeJournal);

        when(aktorConsumer.getAktorId("fnr")).thenReturn("aktorId");
        when(eksisterendeSakService.finnEksisterendeSak("aktorId", "orgnummer")).thenReturn(null);

        when(behandlendeEnhetConsumer.hentBehandlendeEnhet("fnr")).thenReturn("enhet");
        when(behandlendeEnhetConsumer.hentGeografiskTilknytning("fnr")).thenReturn(GeografiskTilknytningData.builder().geografiskTilknytning("tilknytning").diskresjonskode("").build());
    }

    private InngaaendeJournal inngaaendeJournal(String arkivId) {
        return InngaaendeJournal
                .builder()
                .status(MIDLERTIDIG)
                .dokumentId(arkivId)
                .build();
    }

    @Test
    public void gjenbrukerSaksIdHvisViFarToOverlappendeInntektsmeldinger() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"));
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"));

        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 1, 1)).tom(LocalDate.of(2019, 1, 16)).build())
                ).getBytes()),
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 1, 2)).tom(LocalDate.of(2019, 1, 16)).build())
                ).getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"));
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"));

        List<InntektsmeldingMeta> inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");
        assertThat(inntektsmeldingMetas.size()).isEqualTo(2);
        assertThat(inntektsmeldingMetas.get(0).getSakId()).isEqualTo(inntektsmeldingMetas.get(1).getSakId());

        verify(behandleSakConsumer, times(1)).opprettSak("fnr");
    }

    @Test
    public void gjenbrukerIkkeSaksIdHvisViFarToInntektsmeldingerSomIkkeOverlapper() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"));
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"));

        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 1, 1)).tom(LocalDate.of(2019, 1, 16)).build())
                ).getBytes()),
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 2, 2)).tom(LocalDate.of(2019, 2, 16)).build())
                ).getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"));
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"));

        List<InntektsmeldingMeta> inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");
        assertThat(inntektsmeldingMetas.size()).isEqualTo(2);
        assertThat(inntektsmeldingMetas.get(0).getSakId()).isNotEqualTo(inntektsmeldingMetas.get(1).getSakId());

        assertThat(inntektsmeldingMetas.get(0).getSakId()).isEqualTo("saksId0");
        assertThat(inntektsmeldingMetas.get(1).getSakId()).isEqualTo("saksId1");

        verify(behandleSakConsumer, times(2)).opprettSak("fnr");
    }

    @Test
    public void brukerSaksIdFraSykeforloepOmViIkkeHarOverlappendeInntektsmelding() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(eksisterendeSakService.finnEksisterendeSak("aktorId", "orgnummer")).thenReturn(null, "syfosak");
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"));
        when(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"));

        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 1, 1)).tom(LocalDate.of(2019, 1, 16)).build())
                ).getBytes()),
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        Collections.singletonList(Periode.builder().fom(LocalDate.of(2019, 2, 2)).tom(LocalDate.of(2019, 2, 16)).build())
                ).getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"));
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"));

        List<InntektsmeldingMeta> inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");
        assertThat(inntektsmeldingMetas.size()).isEqualTo(2);
        assertThat(inntektsmeldingMetas.get(0).getSakId()).isNotEqualTo(inntektsmeldingMetas.get(1).getSakId());

        assertThat(inntektsmeldingMetas.get(0).getSakId()).isEqualTo("saksId0");
        assertThat(inntektsmeldingMetas.get(1).getSakId()).isEqualTo("syfosak");

        verify(behandleSakConsumer, times(1)).opprettSak("fnr");
    }

    @Test
    public void mottarInntektsmeldingUtenArbeidsgiverperioder() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(emptyList()).getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"));

        List<InntektsmeldingMeta> inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");

        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().size()).isEqualTo(0);
    }

    @Test
    public void mottarInntektsmeldingMedFlerePerioder() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiver(
                        asList(Periode.builder().fom(LocalDate.of(2019, 1, 1)).tom(LocalDate.of(2019, 1, 12)).build(),
                                Periode.builder().fom(LocalDate.of(2019, 1, 12)).tom(LocalDate.of(2019, 1, 14)).build())
                ).getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"));

        List<InntektsmeldingMeta> inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");

        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().size()).isEqualTo(2);
        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().get(0).getFom()).isEqualTo(LocalDate.of(2019, 1, 1));
        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().get(0).getTom()).isEqualTo(LocalDate.of(2019, 1, 12));
        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().get(1).getFom()).isEqualTo(LocalDate.of(2019, 1, 12));
        assertThat(inntektsmeldinger.get(0).getArbeidsgiverperioder().get(1).getTom()).isEqualTo(LocalDate.of(2019, 1, 14));
    }

    @Test
    public void mottarInntektsmeldingMedPrivatArbeidsgiver() throws MessageNotWriteableException, HentDokumentSikkerhetsbegrensning, HentDokumentDokumentIkkeFunnet {
        when(journalV2.hentDokument(any())).thenReturn(
                new WSHentDokumentResponse().withDokument(JournalConsumerTest.inntektsmeldingArbeidsgiverPrivat().getBytes())
        );

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"));

        List<InntektsmeldingMeta> inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId");

        assertThat(inntektsmeldinger.get(0).getArbeidsgiverPrivat()).isEqualTo("arbeidsgiverPrivat");
        assertThat(inntektsmeldinger.get(0).getOrgnummer()).isNull();
        assertThat(inntektsmeldinger.get(0).getAktorId()).isEqualTo("aktorId");
    }

    private ActiveMQTextMessage opprettKoemelding(String arkivId) throws MessageNotWriteableException {
        ActiveMQTextMessage message = new ActiveMQTextMessage();
        message.setText("" +
                "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>" +
                "  <ns5:forsendelsesinformasjon xmlns:ns5=\"http://nav.no/melding/virksomhet/dokumentnotifikasjon/v1\" " +
                "    xmlns:ns2=\"http://nav.no/melding/virksomhet/dokumentforsendelse/v1\" " +
                "    xmlns:ns4=\"http://nav.no/dokmot/jms/reply\" " +
                "    xmlns:ns3=\"http://nav.no.dokmot/jms/viderebehandling\">" +
                "  <arkivId>" + arkivId + "</arkivId>" +
                "  <arkivsystem>JOARK</arkivsystem>" +
                "  <tema>SYK</tema>" +
                "  <behandlingstema>ab0061</behandlingstema>" +
                "</ns5:forsendelsesinformasjon>");

        return message;
    }
}

File: src/test/kotlin/no/nav/syfo/consumer/mq/InntektsmeldingConsumerIntegrassjonsTest.kt
package no.nav.syfo.consumer.mq

import any
import kotlinx.coroutines.runBlocking
import no.nav.syfo.LocalApplication
import no.nav.syfo.consumer.rest.OppgaveClient
import no.nav.syfo.consumer.rest.SakClient
import no.nav.syfo.consumer.rest.SakResponse
import no.nav.syfo.consumer.rest.aktor.AktorConsumer
import no.nav.syfo.consumer.ws.BehandlendeEnhetConsumer
import no.nav.syfo.consumer.ws.InngaaendeJournalConsumer
import no.nav.syfo.consumer.ws.JournalConsumerTest
import no.nav.syfo.consumer.ws.OppgavebehandlingConsumer
import no.nav.syfo.domain.GeografiskTilknytningData
import no.nav.syfo.domain.InngaaendeJournal
import no.nav.syfo.domain.JournalStatus
import no.nav.syfo.domain.Periode
import no.nav.syfo.producer.InntektsmeldingProducer
import no.nav.syfo.repository.InntektsmeldingDAO
import no.nav.syfo.service.EksisterendeSakService
import no.nav.syfo.util.Metrikk
import no.nav.tjeneste.virksomhet.behandleinngaaendejournal.v1.binding.BehandleInngaaendeJournalV1
import no.nav.tjeneste.virksomhet.journal.v2.binding.HentDokumentDokumentIkkeFunnet
import no.nav.tjeneste.virksomhet.journal.v2.binding.HentDokumentSikkerhetsbegrensning
import no.nav.tjeneste.virksomhet.journal.v2.binding.JournalV2
import no.nav.tjeneste.virksomhet.journal.v2.meldinger.HentDokumentRequest
import no.nav.tjeneste.virksomhet.journal.v2.meldinger.HentDokumentResponse
import org.apache.activemq.command.ActiveMQTextMessage
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.BDDMockito.given
import org.mockito.Mockito.*
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.TestPropertySource
import org.springframework.test.context.junit4.SpringRunner
import java.time.LocalDate
import java.time.ZonedDateTime
import java.util.Arrays.asList
import java.util.Collections.emptyList
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject
import javax.jms.MessageNotWriteableException

@RunWith(SpringRunner::class)
@SpringBootTest(classes = [LocalApplication::class])
@TestPropertySource("classpath:application-test.properties")
@DirtiesContext
class InntektsmeldingConsumerIntegrassjonsTest {

    @MockBean
    private lateinit var inngaaendeJournalConsumer: InngaaendeJournalConsumer

    @MockBean
    private lateinit var journalV2: JournalV2

    @MockBean
    private lateinit var aktorConsumer: AktorConsumer

    @MockBean
    private lateinit var eksisterendeSakService: EksisterendeSakService

    @MockBean
    private lateinit var behandlendeEnhetConsumer: BehandlendeEnhetConsumer

    @MockBean
    private lateinit var oppgavebehandlingConsumer: OppgavebehandlingConsumer

    @MockBean
    private lateinit var behandleInngaaendeJournalV1: BehandleInngaaendeJournalV1

    @MockBean
    private val metrikk: Metrikk? = null

    @Inject
    private lateinit var inntektsmeldingConsumer: InntektsmeldingConsumer

    @MockBean
    private lateinit var inntektsmeldingProducer: InntektsmeldingProducer

    @Inject
    private lateinit var inntektsmeldingDAO: InntektsmeldingDAO

    @Inject
    private lateinit var jdbcTemplate: JdbcTemplate

    @MockBean
    private lateinit var oppgaveClient : OppgaveClient

    @MockBean
    private lateinit var sakClient : SakClient

    @Before
    fun setup() {
        jdbcTemplate.update("DELETE FROM ARBEIDSGIVERPERIODE")
        jdbcTemplate.update("DELETE FROM INNTEKTSMELDING")

        val inngaaendeJournal = inngaaendeJournal("arkivId")

        runBlocking {
            given(sakClient.opprettSak( any(), any())).willReturn(
                    SakResponse(id = 987, tema = "SYM", aktoerId = "444", applikasjon = "", fagsakNr = "123000", opprettetAv = "meg", opprettetTidspunkt = ZonedDateTime.now(), orgnr = "999888777"),
                    SakResponse(id = 988, tema = "SYM", aktoerId = "444", applikasjon = "", fagsakNr = "123000", opprettetAv = "meg", opprettetTidspunkt = ZonedDateTime.now(), orgnr = "999888777")
            )
        }

        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId")).thenReturn(inngaaendeJournal)

        given(aktorConsumer.getAktorId(ArgumentMatchers.anyString())).willAnswer { "aktorId_for_" + it.getArgument(0) }

        given(eksisterendeSakService.finnEksisterendeSak(ArgumentMatchers.anyString(), any(), any())).willReturn(null)

        `when`(behandlendeEnhetConsumer.hentBehandlendeEnhet(ArgumentMatchers.anyString())).thenReturn("enhet")
        `when`(behandlendeEnhetConsumer.hentGeografiskTilknytning(ArgumentMatchers.anyString())).thenReturn(
                GeografiskTilknytningData(geografiskTilknytning = "tilknytning", diskresjonskode = "")
        )
    }

    private fun inngaaendeJournal(arkivId: String): InngaaendeJournal {
        return InngaaendeJournal(
                dokumentId = arkivId,
                status = JournalStatus.MIDLERTIDIG)
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun gjenbrukerSaksIdHvisViFarToOverlappendeInntektsmeldinger() {
        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"))
        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"))

        val dokumentResponse1 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse1.response = HentDokumentResponse()
        dokumentResponse1.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 1, 16)))
        ).toByteArray()


        val dokumentResponse2 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse2.response = HentDokumentResponse()
        dokumentResponse2.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 1, 2), LocalDate.of(2019, 1, 16)))
        ).toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse1.response,
                dokumentResponse2.response
        )

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"))
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"))

        val inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId_for_fnr")
        assertThat(inntektsmeldingMetas.size).isEqualTo(2)
        assertThat(inntektsmeldingMetas[0].sakId).isEqualTo(inntektsmeldingMetas[1].sakId)

        runBlocking {
            verify<SakClient>(sakClient, times(1)).opprettSak("aktorId_for_fnr", "")
        }
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun gjenbrukerIkkeSaksIdHvisViFarToInntektsmeldingerSomIkkeOverlapper() {
        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"))
        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"))

        val dokumentResponse1 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse1.response = HentDokumentResponse()
        dokumentResponse1.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 1, 16)))
        ).toByteArray()


        val dokumentResponse2 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse2.response = HentDokumentResponse()
        dokumentResponse2.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 2, 2), LocalDate.of(2019, 2, 16)))
        ).toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse1.response,
                dokumentResponse2.response
        )

        given(aktorConsumer.getAktorId(ArgumentMatchers.anyString())).willReturn("777", "777")

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"))
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"))

        val inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("777")
        assertThat(inntektsmeldingMetas.size).isEqualTo(2)
        assertThat(inntektsmeldingMetas[0].sakId).isNotEqualTo(inntektsmeldingMetas[1].sakId)

        assertThat(inntektsmeldingMetas[0].sakId).isEqualTo("987")
        assertThat(inntektsmeldingMetas[1].sakId).isEqualTo("988")

        runBlocking {
            verify<SakClient>(sakClient, times(2)).opprettSak("777", "")
        }
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun brukerSaksIdFraSykeforloepOmViIkkeHarOverlappendeInntektsmelding() {
        given(eksisterendeSakService.finnEksisterendeSak(any(), any(), any())).willReturn(null, "syfosak")

        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId1")).thenReturn(inngaaendeJournal("arkivId1"))
        `when`(inngaaendeJournalConsumer.hentDokumentId("arkivId2")).thenReturn(inngaaendeJournal("arkivId2"))

        val dokumentResponse1 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse1.response = HentDokumentResponse()
        dokumentResponse1.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 1, 16)))
        ).toByteArray()

        val dokumentResponse2 = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse2.response = HentDokumentResponse()
        dokumentResponse2.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(Periode(LocalDate.of(2019, 2, 2), LocalDate.of(2019, 2, 16)))
        ).toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse1.response,
                dokumentResponse2.response
        )
        given(aktorConsumer.getAktorId(ArgumentMatchers.anyString())).willReturn("999", "999")

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId1"))
        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId2"))

        val inntektsmeldingMetas = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("999")
        assertThat(inntektsmeldingMetas.size).isEqualTo(2)
        assertThat(inntektsmeldingMetas[0].sakId).isNotEqualTo(inntektsmeldingMetas[1].sakId)

        assertThat(inntektsmeldingMetas[0].sakId).isEqualTo("987")
        assertThat(inntektsmeldingMetas[1].sakId).isEqualTo("syfosak")

        runBlocking {
            verify<SakClient>(sakClient, times(1)).opprettSak(ArgumentMatchers.anyString(), ArgumentMatchers.anyString())
        }
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun mottarInntektsmeldingUtenArbeidsgiverperioder() {
        val dokumentResponse = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse.response = HentDokumentResponse()
        dokumentResponse.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                emptyList<Periode>()
        ).toByteArray()
        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse.response
        )

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"))

        val inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId_for_fnr")

        assertThat(inntektsmeldinger[0].arbeidsgiverperioder.size).isEqualTo(0)
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun mottarInntektsmeldingMedFlerePerioder() {
        val dokumentResponse = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse.response = HentDokumentResponse()
        dokumentResponse.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                asList(
                        Periode(LocalDate.of(2019, 1, 1), LocalDate.of(2019, 1, 12)),
                        Periode(LocalDate.of(2019, 1, 12), LocalDate.of(2019, 1, 14))
                )
        ).toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(dokumentResponse.response)

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"))

        val inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId_for_fnr")

        assertThat(inntektsmeldinger[0].arbeidsgiverperioder.size).isEqualTo(2)
        assertThat(inntektsmeldinger[0].arbeidsgiverperioder[0].fom).isEqualTo(LocalDate.of(2019, 1, 1))
        assertThat(inntektsmeldinger[0].arbeidsgiverperioder[0].tom).isEqualTo(LocalDate.of(2019, 1, 12))
        assertThat(inntektsmeldinger[0].arbeidsgiverperioder[1].fom).isEqualTo(LocalDate.of(2019, 1, 12))
        assertThat(inntektsmeldinger[0].arbeidsgiverperioder[1].tom).isEqualTo(LocalDate.of(2019, 1, 14))
    }

    @Test
    @Throws(
            MessageNotWriteableException::class,
            HentDokumentSikkerhetsbegrensning::class,
            HentDokumentDokumentIkkeFunnet::class
    )
    fun mottarInntektsmeldingMedPrivatArbeidsgiver() {
        val dokumentResponse = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse.response = HentDokumentResponse()
        dokumentResponse.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiverPrivat().toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse.response
        )

        inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"))

        val inntektsmeldinger = inntektsmeldingDAO.finnBehandledeInntektsmeldinger("aktorId_for_fnr")

        assertThat(inntektsmeldinger[0].arbeidsgiverPrivat).isEqualTo("arbeidsgiverPrivat")
        assertThat(inntektsmeldinger[0].orgnummer).isNull()
        assertThat(inntektsmeldinger[0].aktorId).isEqualTo("aktorId_for_fnr")
    }

    @Throws(MessageNotWriteableException::class)
    private fun opprettKoemelding(arkivId: String): ActiveMQTextMessage {
        val message = ActiveMQTextMessage()
        message.text = "" +
                "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>" +
                "  <ns5:forsendelsesinformasjon xmlns:ns5=\"http://nav.no/melding/virksomhet/dokumentnotifikasjon/v1\" " +
                "    xmlns:ns2=\"http://nav.no/melding/virksomhet/dokumentforsendelse/v1\" " +
                "    xmlns:ns4=\"http://nav.no/dokmot/jms/reply\" " +
                "    xmlns:ns3=\"http://nav.no.dokmot/jms/viderebehandling\">" +
                "  <arkivId>" + arkivId + "</arkivId>" +
                "  <arkivsystem>JOARK</arkivsystem>" +
                "  <tema>SYK</tema>" +
                "  <behandlingstema>ab0061</behandlingstema>" +
                "</ns5:forsendelsesinformasjon>"

        return message
    }

    @Test
    @Throws(Exception::class)
    fun behandlerInntektsmeldingSomEnSakMedVedLikPeriode() {
        val dokumentResponse = no.nav.tjeneste.virksomhet.journal.v2.HentDokumentResponse()
        dokumentResponse.response = HentDokumentResponse()
        dokumentResponse.response.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(
                        Periode(LocalDate.now(), LocalDate.now().plusDays(20))
                )
        ).toByteArray()

        `when`(journalV2.hentDokument(any())).thenReturn(
                dokumentResponse.response
        )

        val numThreads = 16
        produceParallelMessages(numThreads)

        runBlocking {
            verify<SakClient>(sakClient, times(1)).opprettSak(ArgumentMatchers.anyString(), ArgumentMatchers.anyString())
        }
        verify<BehandleInngaaendeJournalV1>(behandleInngaaendeJournalV1, times(numThreads)).ferdigstillJournalfoering(
                any()
        )
    }

    fun build(fnr: AtomicInteger): no.nav.tjeneste.virksomhet.journal.v2.meldinger.HentDokumentResponse {
        val dokumentResponse = no.nav.tjeneste.virksomhet.journal.v2.meldinger.HentDokumentResponse()
        dokumentResponse.dokument = JournalConsumerTest.inntektsmeldingArbeidsgiver(
                listOf(
                        Periode(
                                fom = LocalDate.now(),
                                tom = LocalDate.now().plusDays(20)
                        )
                ),
                "fnr" + fnr.incrementAndGet()
        ).toByteArray()
        return dokumentResponse
    }

    @Test
    @Throws(Exception::class)
    fun behandlerInntektsmeldingerForFlerPersonerSamtidig() {
        val fnr = AtomicInteger()
        given(journalV2.hentDokument(any<HentDokumentRequest>())).willAnswer {
            build(fnr)
        }

        val numThreads = 16
        produceParallelMessages(numThreads)
        runBlocking {
            verify<SakClient>(sakClient, times(numThreads)).opprettSak(ArgumentMatchers.anyString(), ArgumentMatchers.anyString())
        }
        verify<BehandleInngaaendeJournalV1>(behandleInngaaendeJournalV1, times(numThreads)).ferdigstillJournalfoering(
                any()
        )
    }

    @Throws(Exception::class)
    fun produceParallelMessages(numThreads: Int) {
        val countdown = CountDownLatch(numThreads)

        repeat(numThreads) {
            Thread {
                try {
                    inntektsmeldingConsumer.listen(opprettKoemelding("arkivId"))
                } catch (e: MessageNotWriteableException) {
                    e.printStackTrace()
                }

                countdown.countDown()
            }.start()
        }
        countdown.await()
    }
}
